HPCC-33921: Add event filtering to evtool commands

timothyklemm opened the pull request:
- The sim command does not filter.
- All other commands filter on events using event names and event contexts
  (hard coded associations of related events, such as 'Index' or 'Dali').
- All other commands filter on attributes using attribute names.
  - Boolean attribute filters use strToBool to recognize true or false.
  - Numeric attribute filters can use single values or ranges, where ranges
    may be closed on one or both ends.
  - Timestamp attribute filters extend numeric filters to recognize timestamp
    text representations supported by CDateTime.
  - Text attribute filters support exact match, wildcard match, and regular
    expression match.
  - FileId filters extend numeric filters to also accept text patterns.
    Numeric filters apply to the attribute value itself. Text filters join
    with a corresponding FileInformation meta event and apply the filter to
    the Path attribute.

Refactored command and operation class hierarchies to facilitate reuse of
filter request and application.

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

Commit:
HPCC-33921: Add event filtering to evtool commands

- The sim command does not filter.
- All other commands filter on events using event names and event contexts
  (hard coded associations of related events, such as 'Index' or 'Dali').
- All other command filter on attributes using attribute names.
  - Boolean attribute filters use strToBool to recognize true or false.
  - Numeric attribute filters can use single values or ranges, where ranges
    may be closed on one or both ends.
  - Timestamp attribute filters extend numeric filters to recognize timestamp
    text representations supported by CDateTime.
  - Text attribute filters support exact match, wildcard match, and regular
    expression match.
  - FileId filters extend numeric filters to also accept text patterns.
    Numeric filters apply to the attribute value itself. Text filters join
    with a corresponding FileInformation meta event and apply the filter to
    the Path attribute.

Refactored command and operation class hierarchies to facilitate reuse of
filter request and application.

Signed-off-by: Tim Klemm <Tim.Klemm@lexisnexisrisk.com>

ghalliday commented on tools/evtool/evtindex.hpp:
```diff
@@ -114,5 +114,45 @@ class CStreamingBucketVisitor : public CInterfaceOf<IBucketVisitor>
+// `event_consuming_op_t` must implement the following extended interface:
+// - `bool filterFileId(const char* fileId)`: adds a file ID filter to the operation
+// - `bool filterFilePath(const char* path)`: adds a file path filter to the operation
+template <typename event_consuming_op_t>
```
I am not a big fan of using template classes like this for a couple of reasons
i) the code is hard to follow
ii) The code gets replicated into every derived class, bumping up the generated code size.

An alternative is to add the functionality in the base class, with a flag that is checked to see if it is applicable.
Something to discuss the relative merits.

ghalliday commented on tools/evtool/evtindex.hpp:
```diff
@@ -114,5 +114,45 @@ class CStreamingBucketVisitor : public CInterfaceOf<IBucketVisitor>
     bool arrived{false};
 };
 
+class CIndexEventsOp : public CEventConsumingOp
```
What does this class represent?

ghalliday commented on tools/evtool/evtool.hpp:
```diff
@@ -29,55 +30,191 @@ class CEvToolCommand : public CInterfaceOf<IEvToolCommand>
+// template parameter of `TEventConsumingCommand`.
+//
+// Subclasses are expected to add support for additional filters and options.
+class CEventConsumingOp : public CEventFileOp
```
Is an Op really a command from the tools point of view?

ghalliday commented on system/jlib/jevent.cpp:
```diff
@@ -1754,6 +1758,913 @@ IEventVisitor* createDumpCSVEventVisitor(IBufferedSerialOutputStream& out)
+
+    virtual Continuation visitEvent(EventType type) override
+    {
+        if (!available)
```
I don't think this is worth optimizing, and the extra complication it adds to the logic - having to set the flag up and then test it.

For all the tests I would expect logic like this:
i) at start of event clear flags to indicate a filter has been applied.
ii) When an attribute is received, check any associated filters, and add it to a list if the items is not being removed.
iii) When an event is received, check if all filters were applied.  If not, then optionally reject the event (depending on whether the filter is required or not).
iv) Pass the event through to the visitor.

I also do not think reusing the visitor interface for the filters element is the correct approach.  The only functions you need are to check if a particular attribute matches or not.  The interface could be much simpler.


My naive implementation would be:
For each eventattr have a filter that must be matched.  The filter only receives the value to check if it is valid.
Filename filters need special casing, but I think I would locate that logic in the filter class, not in the filters.
Depending on how the implementation fell out I might require some of the filters for fields to match a particular form.

I'll add some extra comments on other items below.

ghalliday commented on system/jlib/jevent.cpp:
```diff
@@ -1754,6 +1758,913 @@ IEventVisitor* createDumpCSVEventVisitor(IBufferedSerialOutputStream& out)
+            }
+        case EvAttrEventTimeOffset:
+            {
+                __uint64 micros = (base + value ) / 1000;
```
This should be implemented as a simple range test (similar to any other numeric field).

The values in the list should be adjusted when the EvAttrRecordedTimestamp field is seen.  As an alternative I think on balance the timestamp at the start of the file should be added to all time offsets, and passed through to the visitor as a real timestamp.  (Really the fact the an offset is stored in the file is an implementation detail that I don't think should be visible to the outside world).
That then means this filter becomes a simple numeric range filter.

ghalliday commented on system/jlib/jevent.cpp:
```diff
@@ -1754,6 +1758,913 @@ IEventVisitor* createDumpCSVEventVisitor(IBufferedSerialOutputStream& out)
+    bool accepted;
+};
+
+class CLogicalEventVisitFilterTerm : public CInterfaceOf<ILogicalEventVisitFilterTerm>
```
This would almost certainly be simpler with a single class handling and and or with a flag.  The virtual taking a std::function is really not very efficient.  I am not sure if you really need either of these.
(For discission.)

ghalliday commented on system/jlib/jevent.cpp:
```diff
@@ -1754,6 +1758,913 @@ IEventVisitor* createDumpCSVEventVisitor(IBufferedSerialOutputStream& out)
+            }
+        };
+
+        struct EventVisit : public Visit
```
The pending list is more complex than I would expect (e.g. files, and events).  Probably worth revisiting in the light of other changes.

E.g. I would probably start with a vector of (attribute, type, value-union)

ghalliday reviewed (commented):
My initial impression is that this is more complex, and much larger than I expected.  It would be better to release the changes incrementally so that the underlying design can be discussed first - implementing one or two filters and iterating.
 
I think it would be worth discussing
- allow the filter instance to be created early, expose functions for adding the filters, and then set the visitor it is applied to later.  That should allow the interfaces to be simplified.
- simplifying the filter implementation classes.
- Change the way timestamps are passed through (a separate PR) and allow that to simplify timestamp filtering.

Add documentation on what filters on each of the fields mean if the event does that have that attribute.  Probably add a separate boolean for each attribute, whether it is required to be present in the event.  Generally setting a filter would also require the field, but branch/leaf and filename filters introduce some weird exceptions.

timothyklemm commented:
@ghalliday I squashed and rebased the existing PR rather than closing it and creating a new one. I completed more spot checks, including path (exact match, wildcard match, and regex match), file ID (xref with file information path), and timestamp (numeric and text ranges).

ghalliday commented on tools/evtool/evtool.hpp:
```diff
@@ -29,55 +30,268 @@ class CEvToolCommand : public CInterfaceOf<IEvToolCommand>
+                                - EventThreadId: numeric
+                                - EventStackTrace: string
+                                - DataSize: numeric
+                              A limited number of attributes are unfulterable
```
typo: filterable

ghalliday replied:
I'm not sure there is any benefit in including this information though - the user won't even know about them.

timothyklemm replied:
jevent.hpp defines the attributes in the enumeration. Every visitor will see them, even if they don't disclose them. Awareness is possible, if not likely, and I would prefer to address them up-front. Having said that...

RecordedTimestamp has limited value now that it is combined with the event offset to make EventTimestamp. There might be value when no events are visited. Not sure seeing a gap between start of recording and the first event. I could be convinced to remove the attribute entirely.

RecordedOption had value with the idea that a visitor could react to what is/isn't in the file, before visiting event attributes. The same information can be inferred from event data, if we ever want to. I could be convinced to remove the attribute entirely.

If the attributes are removed, this documentation will necessarily be removed with them. If they remain, I think the documentation should remain.

ghalliday commented on system/jlib/jevent.cpp:
```diff
@@ -146,13 +150,68 @@ static constexpr EventAttrInformation attrInformation[] = {
+    case EventIndexLoad:
+    case EventIndexEviction:
+    case MetaFileInformation:
+        return EventCtxIndex;
```
I would add a field to the eventInformation and use that to implement - then the information is all in one place.

ghalliday commented on system/jlib/jevent.cpp:
```diff
@@ -1757,6 +1821,513 @@ IEventVisitor* createDumpCSVEventVisitor(IBufferedSerialOutputStream& out)
+            {
+                if (tmp.compare(pattern) == 0)
+                    return true;
+                if (isWildString(pattern.c_str()))
```
This will match "abc.*".  I am not sure wildcards add anything - regex should be sufficient.

ghalliday commented on system/jlib/jevent.cpp:
```diff
@@ -1757,6 +1821,513 @@ IEventVisitor* createDumpCSVEventVisitor(IBufferedSerialOutputStream& out)
+
+        bool matches(EventType evt, bool value) const override
+        {
+            return (value && acceptTrue) || (!value && acceptFalse);
```
picky: Maybe clearer as (value ? acceptTrue : acceptFalse)

You could also use an array[2]: acceptValue[value]
 - bool conversion to int is defined.

ghalliday commented on system/jlib/jevent.cpp:
```diff
@@ -1757,6 +1821,513 @@ IEventVisitor* createDumpCSVEventVisitor(IBufferedSerialOutputStream& out)
+        {
+            if (first < last)
+                return accepted.insert(std::make_pair(first, last)).second;
+            return accepted.insert(std::make_pair(last, first)).second;
```
return false instead of reversing the arguments?

ghalliday commented on system/jlib/jevent.cpp:
```diff
@@ -1757,6 +1821,513 @@ IEventVisitor* createDumpCSVEventVisitor(IBufferedSerialOutputStream& out)
+
+        bool acceptToken(const char* token) override
+        {
+            char* tmp = nullptr;
```
minor: next is a clearer name than tmp

ghalliday commented on system/jlib/jevent.cpp:
```diff
@@ -1757,6 +1821,513 @@ IEventVisitor* createDumpCSVEventVisitor(IBufferedSerialOutputStream& out)
+            }
+            else
+            {
+                __uint64 first = strtoull(token, &tmp, 0);
```
Code here could be simplified, but I'm not sure it is worth revisiting.   (e.g. Avoid most of the special casing, use variables for low and high, return false on error, and have a single call to add the range.)

ghalliday commented on system/jlib/jevent.cpp:
```diff
@@ -1757,6 +1821,513 @@ IEventVisitor* createDumpCSVEventVisitor(IBufferedSerialOutputStream& out)
+            return acceptRange(first, last);
+        }
+
+        void trim(const char*& token)
```
better named as skipSpaces - because trim generally means removed them from the end.  Can be a static/non member function.

ghalliday commented on system/jlib/jevent.cpp:
```diff
@@ -1757,6 +1821,513 @@ IEventVisitor* createDumpCSVEventVisitor(IBufferedSerialOutputStream& out)
+    }
+
+protected:
+    using AcceptedEvents = std::set<EventType>;
```
std::unordered_set may be better?

ghalliday commented on system/jlib/jevent.cpp:
```diff
@@ -1757,6 +1821,513 @@ IEventVisitor* createDumpCSVEventVisitor(IBufferedSerialOutputStream& out)
+    public:
+        bool matches(const char* value) const
+        {
+            std::string tmp(value);
```
Avoid cloning the string - unless it is necessary. (e.g. use strcmp)

ghalliday commented on system/jlib/jevent.cpp:
```diff
@@ -1757,6 +1821,513 @@ IEventVisitor* createDumpCSVEventVisitor(IBufferedSerialOutputStream& out)
+        {
+            for (const std::pair<__uint64, __uint64>& range : accepted)
+            {
+                if (value > range.second)
```
Should have a comment why this is here.  The code would read more easily if this followed the range test.
I am not sure if is worth including, and/or whether it is worth also having a unordered set of exact matches..  Leave in for now.

timothyklemm replied:
Will remove. Adding a comment made me think about an assumption that has to be made for this to be valid. There is no logic in acceptToken to prevent overlapping ranges. Exceeding the upper bound of one range does not mean there isn't a subsequent range that will match.

In normal use, multiple ranges will probably be rare. Enough ranges to make the attempted optimization significant is unlikely to ever happen.

ghalliday reviewed (commented):
@timothyklemm generally looks good, a few scattered comments.

The PR is too big to review effectively though - it is much better to add the functionality in stages.

ghalliday commented on system/jlib/jevent.hpp:
```diff
@@ -285,11 +297,56 @@ interface IEventPTreeCreator : extends IInterface
+    //   or will cancel each other out.
+    //
+    // Supported special cases include:
+    // - Timestammps may be given using a standard date/time format, such as "2025-01-01T00:00:00".
```
typo: timestamps

ghalliday commented on system/jlib/jevent.cpp:
```diff
@@ -1833,30 +1815,12 @@ class CEventFilter : public CInterfaceOf<IEventFilter>
-                if (tmp.compare(pattern) == 0)
+                if (streq(value, pattern.c_str()))
+                    return true;
+                if (isWildString(pattern.c_str()) && WildMatch(value, pattern.c_str(), false))
```
minor: Save a boolean isWild in the constructor to avoid retesting each time.

ghalliday commented on system/jlib/jevent.cpp:
```diff
@@ -1903,21 +1867,18 @@ class CEventFilter : public CInterfaceOf<IEventFilter>
-                acceptTrue = true;
-            if (!value && !acceptFalse)
-                acceptFalse = true;
+            if (!accepted[value])
```
nitpick: Simpler to unconditionally assign.

ghalliday reviewed (commented):
@timothyklemm looks close.  A couple of final comments.

ghalliday reviewed (approved):
@timothyklemm please squash

timothyklemm commented:
@ghalliday squashed.
